mungegithub/github/github.go

```diff
@@ -38,28 +38,90 @@ import (
 
 const (
 	// stolen from https://groups.google.com/forum/#!msg/golang-nuts/a9PitPAHSSU/ziQw1-QHw3EJ
-	maxInt = int(^uint(0) >> 1)
+	maxInt          = int(^uint(0) >> 1)
+	tokenLimit      = 500 // How many github api tokens to not use
+	asyncTokenLimit = 400 // How many github api tokens to not use for 'asyc' calls
 )
 
-type rateLimitRoundTripper struct {
+type callLimitRoundTripper struct {
+	sync.Mutex
+	delegate  http.RoundTripper
+	remaining int
+	resetTime time.Time
+}
+
+func (c *callLimitRoundTripper) getTokenExcept(remaining int) {
+	c.Lock()
+	if c.remaining > remaining {
+		c.remaining--
+		c.Unlock()
+		return
+	}
+	resetTime := c.resetTime
+	c.Unlock()
+	sleepTime := resetTime.Sub(time.Now()) + (1 * time.Minute)
+	glog.Errorf("Ran out of github API tokens. Sleeping for %v minutes", sleepTime.Minutes())
+	// negative duration is fine, it means we are past the github api reset and we won't sleep
+	time.Sleep(sleepTime)
+}
+
+func (c *callLimitRoundTripper) getToken() {
+	c.getTokenExcept(tokenLimit)
+}
+
+func (c *callLimitRoundTripper) getAsyncToken() {
+	c.getTokenExcept(asyncTokenLimit)
+}
+
+func (c *callLimitRoundTripper) putToken() {
+	c.Lock()
+	defer c.Unlock()
+	c.remaining++
+}
+
+func (c *callLimitRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if c.delegate == nil {
+		c.delegate = http.DefaultTransport
+	}
+	// TODO Be smart about which should use getToken and which should use getAsyncToken()
+	c.getToken()
+	resp, err := c.delegate.RoundTrip(req)
+	if resp != nil && resp.Header.Get(httpcache.XFromCache) != "" {
+		c.putToken()
+	}
+	return resp, err
+}
+
+// By default github responds to PR requests with:
+//    Cache-Control:[private, max-age=60, s-maxage=60]
+// Which means the httpcache would not consider anything stale for 60 seconds.
+// However, when we re-check 'PR.mergeable' we need to skip the cache.
+// I considered checking the req.URL.Path and only setting max-age=0 when
+// getting a PR or getting the CombinedStatus, as these are the times we need
+// a super fresh copy. But since all of the other calls are only going to be made
+// once per poll loop the 60 second github freshness doesn't matter. So I can't
+// think of a reason not to just keep this simple and always set max-age=0 on
+// every request.
```

Comment:
discussion why i believe we can start using gregjones/httpcache again (to get the huge speedup)